File: 437. 路径总和 III.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def pathSum(self, root, targetSum):
        """
        :type root: Optional[TreeNode]
        :type targetSum: int
        :rtype: int
        """
        # 类似 560 题
        self.pre_sum_dict = {0: 1}
        self.res = 0

        def dfs(node, pre_sum):
            if not node:
                return
            pre_sum += node.val

            # 查
            if pre_sum - targetSum in self.pre_sum_dict:
                self.res += self.pre_sum_dict[pre_sum - targetSum]

            # 记
            self.pre_sum_dict[pre_sum] = self.pre_sum_dict.get(pre_sum, 0) + 1

            # 继续往下查（不用判断 node.left存不存在，因为开头有 if not node 兜底防守）
            dfs(node.left, pre_sum)
            dfs(node.right, pre_sum)

            # 回溯
            self.pre_sum_dict[pre_sum] -= 1

        dfs(root, 0)
        return self.res

File: 437. 路径总和 III.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, targetSum int) int {
	res := 0
	preSumMap := map[int]int{0: 1} //

	var dfs func(node *TreeNode, preSum int)
	dfs = func(node *TreeNode, preSum int) {
		if node == nil {
			return
		}

		currSum := preSum + node.Val
		if count, ok := preSumMap[currSum-targetSum]; ok {
			res += count
		}

		preSumMap[currSum] += 1
		dfs(node.Left, currSum)
		dfs(node.Right, currSum)
		preSumMap[currSum] -= 1
	}

	dfs(root, 0)
	return res
}
